File: LeetCode/0198/golang/lc0198.go
package lc0198

// Runtime complexity: O(n)
// Auxiliary space: O(n)
func rob(nums []int) int {
	n := len(nums)
	// Handle a corner case - only one house to rob:
	if n == 1 {
		return nums[0]
	}

	// dp[i] is the maximum amount that can be robbed in total from houses[0] to houses[i].
	dp := make([]int, n)
	dp[0] = nums[0]
	dp[1] = max(nums[0], nums[1])
	for i := 2; i < n; i++ {
		dp[i] = max(dp[i-2]+nums[i], dp[i-1])
	}
	return dp[n-1]
}

File: LeetCode/0198/python/lc0198.py
from typing import List

# Runtime complexity: O(n)
# Auxiliary space complexity: O(n)
# Subjective level: medium-
# Solved on: 2026-04-10
class Solution:
    def rob(self, houses: List[int]) -> int:
        NUM_HOUSES = len(houses)

        # update: let's just handle the corner case with an if.
        if NUM_HOUSES == 1:
            return houses[0]

        # dp[i] holds the maximum possible gain from breaking into houses [0..i).
        # This extra zeroth entry is necessary only to make this algo work around
        # the case of only one house being present on the street.
        # otherwise we could go with dp[i] == gain(houses[0..i]), and len(dp) == len(houses).
        dp = [0] * NUM_HOUSES
        dp[0] = houses[0]
        dp[1] = max(houses[0], houses[1])
        for i in range(2, NUM_HOUSES):
            dp[i] = max(dp[i-2] + houses[i], dp[i-1])

        return dp[-1]
